include/materials/material_utility.tcc

```diff
@@ -0,0 +1,221 @@
+//! Compute mean stress p (compression positive)
+inline const double mpm::material::p(
+    const Eigen::Matrix<double, 6, 1>& stress) {
+  return (-1. / 3. * (stress(0) + stress(1) + stress(2)));
+}
+
+//! Compute deviatoric stress
+inline const Eigen::Matrix<double, 6, 1> mpm::material::deviatoric_stress(
+    const Eigen::Matrix<double, 6, 1>& stress) {
+
+  // Compute mean p in tension positive
+  const double p = -1.0 * mpm::material::p(stress);
+
+  // Compute deviatoric by subtracting volumetric part
+  Eigen::Matrix<double, 6, 1> deviatoric_stress = stress;
+  for (unsigned i = 0; i < 3; ++i) deviatoric_stress(i) -= p;
+
+  return deviatoric_stress;
+}
+
+//! Compute J2 invariant
+inline const double mpm::material::j2(
+    const Eigen::Matrix<double, 6, 1>& stress) {
+
+  const double j2 = (std::pow((stress(0) - stress(1)), 2) +
+                     std::pow((stress(1) - stress(2)), 2) +
+                     std::pow((stress(0) - stress(2)), 2)) /
+                        6.0 +
+                    std::pow(stress(3), 2) + std::pow(stress(4), 2) +
+                    std::pow(stress(5), 2);
+
+  return j2;
+}
+
+//! Compute J3 invariant
+inline const double mpm::material::j3(
+    const Eigen::Matrix<double, 6, 1>& stress) {
+
+  // Compute deviatoric stress
+  Eigen::Matrix<double, 6, 1> deviatoric_stress =
+      mpm::material::deviatoric_stress(stress);
+
+  // Compute J3
+  const double j3 =
+      (deviatoric_stress(0) * deviatoric_stress(1) * deviatoric_stress(2)) -
+      (deviatoric_stress(2) * std::pow(deviatoric_stress(3), 2)) +
+      ((2 * deviatoric_stress(3) * deviatoric_stress(4) *
+        deviatoric_stress(5)) -
+       (deviatoric_stress(0) * std::pow(deviatoric_stress(4), 2)) -
+       (deviatoric_stress(1) * std::pow(deviatoric_stress(5), 2)));
+
+  return j3;
+}
+
+//! Compute deviatoric q
+inline const double mpm::material::q(
+    const Eigen::Matrix<double, 6, 1>& stress) {
+
+  // Compute J2 from
+  const double j2 = mpm::material::j2(stress);
+
+  return (std::sqrt(3 * j2));
```

Comment:
Could we just write out a formula instead of a function call?

Reply:
That defeats the purpose of having this PR right, where we want to reduce repetition of codes, and easier to maintain.

Reply:
function calls are expensive, and this is a single line function for J2, if this gets called a lot, it's a lot more expensive!

Reply:
Another way is to clump the different functions together, so all the other functions that need j2 would be in the same call. I feel if we repeat the single line function, it would be hard to maintain.

Reply:
I am still on the fence about this. Does it mean that we have to copy the function (one line) for both j2 and j3?